src/base/emu-i386/simx86/protmode.c

```diff
@@ -390,3 +391,35 @@ int emu_ldt_write(unsigned char *paddr, uint32_t op, int len)
 	return 1;
 }
 
+void emu_check_read_pagefault(dosaddr_t addr, int len)
+{
+	int attr;
+	if (addr + len > LOWMEM_SIZE + HMASIZE &&
+	    !is_hardware_ram(addr, len) &&
+	    (attr = dpmi_get_page_attribute(addr, len)) == 0) {
+		/* trigger an exception in DPMI */
+		TheCPU.err = EXCP0E_PAGE;
+		/* uncommitted page is never "present" */
+		TheCPU.scp_err = 4;
+		TheCPU.cr2 = addr;
+		longjmp(jmp_env, 0);
+	}
+}
+
+int emu_check_write_pagefault(dosaddr_t addr, uint32_t op, int len)
+{
+	int attr;
+	if (addr + len > LOWMEM_SIZE + HMASIZE &&
+	    !is_hardware_ram(addr, len) &&
+	    (attr = dpmi_get_page_attribute(addr, len)) < 9) {
+		if (emu_ldt_write(MEM_BASE32(addr), op, len))
```

Comment:
Why move ldt write here, and leave vga write there?

Reply:
Because the LDT is in properly allocated shared DPMI memory with an attribute set (of 0x82 to be precise), unlike VGA.